src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/mha_single_token.cpp

```diff
@@ -662,19 +794,102 @@ static void mha_single_token_kernel(const ov::intel_cpu::PlainTensor& query,
     }
 #endif
 
+    // TODO: refactor to seperate files
     if (is_pagedattn) {
-        parallel_for3d_dynamic(B, h_group_num, kv_len, [&](size_t b, size_t h_group, size_t pk) {
+        // if present_key is true, it means q*k is already computed in the caller
+        if (present_key) {
+            parallel_for3d_dynamic(B, h_group_num, kv_len, [&](size_t b, size_t h_group, size_t pk) {
+                auto context_len = static_cast<size_t>(context_lens.ptr<int32_t>()[b]);
+                // kv_len must be valid
+                if (pk < context_len) {
+                    auto block_number = beams.ptr<int32_t>(b)[pk / block_size];
+                    auto block_offset = pk % block_size;
+
+                    for (size_t pq = 0; pq < q_len; pq++) {
+                        for (size_t h = h_group * h_each_group_len; h < (h_group + 1) * h_each_group_len; h++) {
+                            buf_attn_w.ptr<float>(b, h, pq)[pk] =
+                                    dot_product(query.ptr<T>(b, h, pq), present_key.ptr<T2>(block_number, h_group, block_offset),
+                                        S, nullptr, nullptr, nullptr);
+                        }
+                    }
+                }
+            });
+        }
+
+        parallel_for3d_dynamic(B, H, q_len, [&](size_t b, size_t h, size_t pq) {
+            auto cur_kv_len = static_cast<size_t>(context_lens.ptr<int32_t>()[b]);
+            auto ncausal = cur_kv_len;
+            // apply attention mask & sofmax
+            float* alibi_ptr = alibi_mask ? &alibi_mask.at<float>({b, h, pq, 0}, true) : nullptr;
+            uint8_t* attn_mask_ptr = nullptr;
+            auto attn_mask_prec = attention_mask.get_precision();
+            if (attention_mask)
+                attn_mask_ptr = reinterpret_cast<uint8_t*>(&attention_mask.at<T>({b, h, pq, 0}, true));
+            uint8_t* cmask_ptr = causal_mask ? &causal_mask.at<uint8_t>({b, h, pq, 0}, true) : nullptr;
+            attn_softmax_kernel(buf_attn_w.ptr<float>(b, h, pq),
+                                buf_attn_w.ptr<float>(b, h, pq),
+                                d_scale,
+                                alibi_ptr,
+                                attn_mask_ptr,
+                                cmask_ptr,
+                                select_nfltmax_at_0,
+                                ncausal,
+                                cur_kv_len,
+                                attn_mask_prec,
+                                ov::element::f32);
+        });
+
+        // attn_w * V
+        // there are enough works for each thread
+        if (B >= static_cast<size_t>(nthr)) {
+            buf_attn_score.resize<float>({static_cast<size_t>(nthr), q_len, h_each_group_len, S});
+            parallel_for2d_dynamic(B, h_group_num, [&](size_t b, size_t h_group) {
```

Comment:
as we discussed, due to different context length for each work item in current batch, it may benefit performance if we sort work item's length into descendent order, so the longer & slower item can be parallel with the shorter & faster items, let's try that in next release!